File: scripts/content_pipeline/hash_utils.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any

SHA256_LEGACY_CRLF_V1 = "sha256-legacy-crlf-v1"
SHA256_CANONICAL_JSON_V1 = "sha256-canonical-json-v1"


def sha256_raw(path: Path) -> str:
    digest = hashlib.sha256()

    with path.open("rb") as handle:
        for chunk in iter(
            lambda: handle.read(1024 * 1024),
            b"",
        ):
            digest.update(chunk)

    return digest.hexdigest()


def normalize_legacy_crlf_text(text: str) -> str:
    return (
        text.replace("\r\n", "\n")
        .replace("\r", "\n")
        .replace("\n", "\r\n")
    )


# Immutable PR-0041 through PR-0044 evidence was recorded from
# CRLF-normalized working-tree bytes. Keep this compatibility helper scoped to
# those historical evidence fields instead of rewriting substantive artifacts.
def sha256_legacy_crlf(path: Path) -> str:
    return hashlib.sha256(
        normalize_legacy_crlf_text(
            path.read_text(encoding="utf-8")
        ).encode("utf-8")
    ).hexdigest()


def normalize_json_number(value: Any) -> Any:
    if isinstance(value, float) and value.is_integer():
        return int(value)

    return value


def canonicalize_json(value: Any) -> Any:
    if isinstance(value, dict):
        return {
            key: canonicalize_json(value[key])
            for key in sorted(value)
        }

    if isinstance(value, list):
        return [
            canonicalize_json(item)
            for item in value
        ]

    return normalize_json_number(value)


def canonical_json_bytes(value: Any) -> bytes:
    return json.dumps(
        canonicalize_json(value),
        ensure_ascii=False,
        separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")


def canonical_json_sha256(path: Path) -> str:
    return hashlib.sha256(
        canonical_json_bytes(
            json.loads(
                path.read_text(encoding="utf-8")
            )
        )
    ).hexdigest()

File: scripts/content_pipeline/hash_utils.mjs
import {
  createHash,
} from 'node:crypto'
import {
  readFile,
} from 'node:fs/promises'

export const HASH_ALGORITHMS = {
  rawSha256: 'sha256-raw-v1',
  legacyCrlfSha256: 'sha256-legacy-crlf-v1',
  canonicalJsonSha256: 'sha256-canonical-json-v1',
}

export const sha256Bytes = (bytes) =>
  createHash('sha256')
    .update(bytes)
    .digest('hex')

export const sha256Raw = async (filePath) =>
  sha256Bytes(await readFile(filePath))

export const normalizeLegacyCrlfText = (text) =>
  text
    .replace(/\r\n/g, '\n')
    .replace(/\r/g, '\n')
    .replace(/\n/g, '\r\n')

// Immutable PR-0041 through PR-0044 evidence was recorded from CRLF-normalized
// working-tree bytes. Keep this compatibility helper scoped to those historical
// evidence fields instead of rewriting their substantive artifacts.
export const sha256LegacyCrlfFromText = (text) =>
  sha256Bytes(
    Buffer.from(
      normalizeLegacyCrlfText(text),
      'utf8',
    ),
  )

export const sha256LegacyCrlf = async (filePath) =>
  sha256LegacyCrlfFromText(
    await readFile(filePath, 'utf8'),
  )

export const canonicalizeJson = (value) => {
  if (Array.isArray(value)) {
    return value.map(canonicalizeJson)
  }

  if (
    value !== null &&
    typeof value === 'object'
  ) {
    return Object.fromEntries(
      Object.keys(value)
        .sort()
        .map((key) => [
          key,
          canonicalizeJson(value[key]),
        ]),
    )
  }

  return value
}

export const canonicalJsonString = (value) =>
  JSON.stringify(canonicalizeJson(value))

export const canonicalJsonSha256FromValue = (value) =>
  createHash('sha256')
    .update(canonicalJsonString(value), 'utf8')
    .digest('hex')

export const canonicalJsonSha256 = async (filePath) =>
  canonicalJsonSha256FromValue(
    JSON.parse(
      await readFile(filePath, 'utf8'),
    ),
  )
